WalletWasabi/WabiSabi/Models/Decomposition/CombinationsOfASize.cs

```diff
@@ -0,0 +1,180 @@
+{
+	// Represents an ordered set of possible decompositions (multisets of output
+	// values) with a specific number of outputs, and support for efficient
+	// range queries based on some constraints used to efficiently allow
+	// extending to the next size up.
+	internal record CombinationsOfASize
+	{
+		public CombinationsOfASize(ImmutableArray<Money> values, Money maximumTotalValue, Money minimumTotalValue)
+		{
+			Values = values.OrderByDescending(x => x).ToImmutableArray();
+
+			ByTotalValue = Values
+				.Select(x => new Decomposition(x))
+				.ToImmutableArray();
+
+			MaximumTotalValue = maximumTotalValue;
+
+			MinimumTotalValue = minimumTotalValue;
+		}
+
+		public ImmutableArray<Decomposition> ByTotalValue { get; private init; }
+
+		public ImmutableArray<Money> Values { get; }
+
+		public int Count => ByTotalValue.Length;
+
+		public int Size => ByTotalValue[0].Outputs.Length;
+
+		public Money MaximumTotalValue { get; private init; }
+
+		public Money MinimumTotalValue { get; private init; }
+
+		// Returns the set of decompositions of the next size up by extending
+		// the decompositions of the current set with each of the base
+		// denominations.
+		//
+		// This is done by generating one set for each base value, which retains
+		// the order, and merging the resulting sets to preserve the global
+		// ordering.
+		internal CombinationsOfASize Extend(Money maximumTotalValue, Money minimumTotalValue)
+			=> this with
+			{
+				MaximumTotalValue = Math.Min(MaximumTotalValue, maximumTotalValue),
+				MinimumTotalValue = Math.Max(MinimumTotalValue, minimumTotalValue),
+
+				ByTotalValue = Values
+				.Select(x => Extend(x, maximumTotalValue, minimumTotalValue))
+				.Aggregate(PossibleDecompositions.MergeDescending)
+				.ToImmutableArray(),
+			};
+
+		// Extend a set of decompositions with a specific output value,
+		// producing combinations that sum to between MinimumTotalValue and
+		// MaximumTotalValue.
+		//
+		// The decompositions in the array are of a uniform size, and ordered
+		// both lexicgraphically (the values of the individual outputs) and by
+		// total value, both descending. Since these values are extended using
+		// additional value, the ordering is preserved.
+		//
+		// First we prune by total value of each decomposition, based on the
+		// target range.
+		//
+		// Then we prune lexicographically, ignoring decompositions that
+		// begin with a higher value.
+		//
+		// Finally we filter, leaving only decompositions which terminate in
+		// a larger value. We can't prune because the last This ensures that
+		// all decompositions are unique.
+		private IEnumerable<Decomposition> Extend(Money additionalOutput, Money maximumTotalValue, Money minimumTotalValue)
+			=> Prune(maximumTotalValue - additionalOutput, minimumTotalValue - additionalOutput, additionalOutput) // FindIndex can handle negative values
+			.Where(x => additionalOutput <= x.Outputs.Last())
```

Comment:
What guarantees that there is always a `last` element? Can the `Outputs` be empty?

Reply:
The combinations start by making singleton decompositions with each of the values, and then extending those repeatedly, so not normally.

It could technically, if you construct a decomposition made from an empty array. It used to not be possible because that was only a private constructor